src/shared/tpm2-util.c

```diff
@@ -229,11 +242,26 @@ int tpm2_context_new(const char *device, Tpm2Context **ret_context) {
         return 0;
 }
 
+/* bug in tpm2-tss: https://github.com/tpm2-software/tpm2-tss/issues/2553 */
+#define TPM2_HR_PERSISTENT_GOOD        (((uint32_t)TPM2_HT_PERSISTENT) << TPM2_HR_SHIFT)
+
 static void tpm2_handle_flush(ESYS_CONTEXT *esys_context, ESYS_TR esys_handle) {
         if (!esys_context || esys_handle == ESYS_TR_NONE)
                 return;
 
-        TSS2_RC rc = sym_Esys_FlushContext(esys_context, esys_handle);
+        /* TPM handles are in the ESYS_TR cache, so this is fast and
```

Comment:
let's keep this logic out of tpm2_handle_flush, this function should be left to *only* flush the handle. checking if it really should be flushed or not can go into tpm2_handle_free, or possibly even higher up

Reply:
> this function should be left to only flush the handle

This function already checks if the handle is `ESYS_TR_NONE` and skips flushing it, so checking if it's persistent or not fit the bill. But since that function has to go away and I have to add something to Tpm2Handle to decide if to flush or not, it has to move up a level.